### PR TITLE
Dataset refactor

### PR DESCRIPTION
**Pull request by jindrahelcl**
This PR introduces a new way how to construct a dataset:

```ini
[train_data]
series=["source", "target", "source_prep", "computed"]
data=["/path/src", ("/path/tgt", <reader>), (<preprocessor>, "source"), <dataset_level_prep>]
outputs=[("target", "/path/out/tgt"), ("computed", "/path/out/cmp", <writer>)]
shuffled=True
buffer_size=1024
```

Details:
* Laziness of a dataset is now captured just in an object's field, not as a subclass
* In lazy dataset, the `buffer_size` determines how much of the data is being pre-fetched. In non-lazy dataset, this is always the size of the data.
* Shuffling can be done both on lazy and non-lazy datasets (lazy dataset will shuffle the buffer)
* `buffer_size` should be bigger than `batch_size` to avoid a warning.
* Writer objects are introduced. A special writer is `neuralmonkey.writers.auto.AutoWriter`, which preserves the functionality from the old codebase, which automatically selects the suitable writer given the type of the data being outputted.
* The old `load_dataset_from_files` function is re-written to work with new dataset to ensure backward compatibility, but a deprecation notice is logged whenever it is used.
* Small refactor was done regarding the sharing of the dataset batching. Now the batching is done exclusively in `learning_utils`, not in the TF manager's `execute` method. 
* Dataset series are now iterators, with the `dataset.get_series()` method returning a fresh iterator every time is done. Internally, the dataset stores factory functions which are called to create a new data iterator for each series.
* A tiny ugly aspect of this is the type resolution done on the `sources` argument. Originally, the `<preprocessor>` would follow the `"source"`  in the tuple, but this would clash with how the file readers are specified (`Tuple[str, Callable]`) so the series-level preprocessor had to become a `Tuple[Callable, str]`

### CODE REVIEW

**Comment by varisd** [comment]
Why is the (<preprocessor>, "source") in this order and not ("source", <preprocessor>)?
I think that it would be clearer to have ('input_file/input_series', 'processor/reader')

**Comment by jindrahelcl** [comment]
Because both reader and processor are `Callable`s and their types can overlap. (Readers are `Callable[[List[str]], Iterator[Any]]`, and preprocessors are `Callable[[List[Any], List[Any]]`, so there is overlap, for example for `Callable[[List[str]], List[str]]`.)

 When you want to recognize which is which according to its type (without having a parent class for `Reader` and `Preprocessor`), you need to switch the ordering, otherwise they are both `Tuple[str, Callable]`.

**Comment by jindrahelcl** [comment]
... and you don't want to start inheriting these things, because it's not that simple because all the invariance/covariance and whatnot.

File: neuralmonkey/dataset.py [hunk 1]
```diff
@@ -0,0 +1,556 @@
...
+                    for s_name, it in self.iterators.items()}
+
+            # Check whether all loaded series have the same length
+            length_dict = {
+                s_name: len(s_data) for s_name, s_data in data.items()}
+            if len(set(length_dict.values())) > 1:
+                raise ValueError("Lengths of data series do not match: {}"
+                                 .format(str(length_dict)))
+
+            self.length = next(iter(length_dict.values()))
```
Comment (varisd) [inline neuralmonkey/dataset.py]: Is the next(iter()) syntax necessary? Shouldn't length_dict.values()[0] be the same?

File: neuralmonkey/dataset.py [hunk 2]
```diff
@@ -0,0 +1,556 @@
...
+            self.iterators = {
+                s_name: lambda n=s_name: iter(data[n])  # type: ignore
+                for s_name in self.iterators}
+
+    def __len__(self) -> int:
+        """Get the length of the dataset.
+
+        Returns:
+            The length of the dataset. If the dataset is lazy, returns None.
```
Comment (varisd) [inline neuralmonkey/dataset.py]: Documentation:
Actually, it throws an exception when lazy.

File: neuralmonkey/dataset.py [hunk 3]
```diff
@@ -0,0 +1,556 @@
...
+        return name in self.iterators
+
+    def get_series(self, name: str) -> Iterator:
+        """Get the data series with a given name.
+
+        Arguments:
+            name: The name of the series to fetch.
+
+        Returns:
+            The data series.
```
Comment (varisd) [inline neuralmonkey/dataset.py]: Documentation:
Isn't the function returning an data_series iterator? I would include it in the documentation if so (for clarity).

File: neuralmonkey/dataset.py (hunk 1 above)
Comment (jindrahelcl) [reply neuralmonkey/dataset.py]:
```ipython
In [1]: a = {"barack": 67, "hussein": 12, "obama": 32}

In [2]: a.values()[0]
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-2-53f4ff1337fa> in <module>()
----> 1 a.values()[0]

TypeError: 'dict_values' object does not support indexing
```

Comment (jindrahelcl) [reply neuralmonkey/dataset.py]: https://stackoverflow.com/questions/3097866/access-an-arbitrary-element-in-a-dictionary-in-python

File: neuralmonkey/dataset.py (hunk 2 above)
Comment (jindrahelcl) [reply neuralmonkey/dataset.py]: good catch..

File: neuralmonkey/dataset.py (hunk 3 above)
Comment (jindrahelcl) [reply neuralmonkey/dataset.py]: Rephrased:

```
Returns:
    A freshly initialized iterator over the data series. 
```

File: neuralmonkey/dataset.py (hunk 1 above)
Comment (varisd) [reply neuralmonkey/dataset.py]: Oh, now I see it.

File: neuralmonkey/writers/plain_text_writer.py~ [hunk 4]
```diff
@@ -0,0 +1,31 @@
+from typing import TypeVar, Iterator, List, Any, Callable
```
Comment (varisd) [inline neuralmonkey/writers/plain_text_writer.py~]: Again, the .py~ file should probably not be included in the commit.

File: neuralmonkey/writers/numpy_writer.py~ [hunk 5]
```diff
@@ -0,0 +1,37 @@
+from typing import Iterator, Dict, List
```
Comment (varisd) [inline neuralmonkey/writers/numpy_writer.py~]: The .py~ file should be deleted

File: neuralmonkey/tests/test_dataset.py [hunk 6]
```diff
@@ -52,7 +53,8 @@ def test_glob(self):
                 s_data=[os.path.join(tmp_dir, "abc?"),
                         os.path.join(tmp_dir, "xyz*")])
 
-            self.assertEqual(dataset.get_series("data"), [["a"], ["b"], ["d"]])
+            series_iterator = dataset.get_series("data")
+            self.assertEqual(list(series_iterator), [["a"], ["b"], ["d"]])
```
Comment (varisd) [inline neuralmonkey/tests/test_dataset.py]: It would probably be good to also include following tests:

- check whether the batching is done correctly (ideally with more than one series)
- check whether the shuffling does not break things (this might be tricky)
- check whether the buffering works

File: neuralmonkey/learning_utils.py [hunk 7]
```diff
@@ -390,18 +390,33 @@ def run_on_dataset(tf_manager: TensorFlowManager,
...
+            script_list.append(ex_result)
+
+    # Transpose runner interim results.
+    all_results = [reduce_execution_results(res) for res in batch_results]
 
+    # Convert execution results to dictionary.
     result_data = {runner.output_series: result.outputs
                    for runner, result in zip(runners, all_results)}
 
+    # Run dataset-level preprocessing.
```
Comment (varisd) [inline neuralmonkey/learning_utils.py]: "postprocessing"

File: neuralmonkey/tests/test_dataset.py (hunk 6 above)
Comment (jindrahelcl) [reply neuralmonkey/tests/test_dataset.py]: hotovo

File: neuralmonkey/learning_utils.py (hunk 7 above)
Comment (jindrahelcl) [reply neuralmonkey/learning_utils.py]: dík

File: neuralmonkey/writers/numpy_writer.py~ (hunk 5 above)
Comment (jindrahelcl) [reply neuralmonkey/writers/numpy_writer.py~]: děkuji

File: neuralmonkey/writers/plain_text_writer.py~ (hunk 4 above)
Comment (jindrahelcl) [reply neuralmonkey/writers/plain_text_writer.py~]: dkuji

File: neuralmonkey/dataset.py [hunk 8]
```diff
@@ -0,0 +1,559 @@
...
+            def itergen():
+                return (row[key] for row in rows)
+            return itergen
+
+        # Iterate over the rest of the data until buffer is empty
+        batch_index = 0
+        while buf:
+            # Create the batch
+            name = "{}.batch.{}".format(self.name, batch_index)
+            rows = [buf.popleft() for _ in range(batch_size) if buf]
```
Comment (varisd) [inline neuralmonkey/dataset.py]: It should probably be mentioned in the documentation that it returns batches of size == buffer_size if buffer_size < batch_size

Comment (jindrahelcl) [reply neuralmonkey/dataset.py]: to je bug, kterej nastane v extrémním případě. je tam warning, ale radši to nějak opravim, aby to vždycky dávalo batch_size

Comment (jindrahelcl) [reply neuralmonkey/dataset.py]: Ono je vůbec otázka, jestli to chceme plnit a shufflovat po každý batchi, nebo jestli se tam nemá dát nějakej lower threshold pro velikost bufferu, pod kterej když to vleze, tak se to doplní na buffer_size

Comment (jindrahelcl) [reply neuralmonkey/dataset.py]: Taky shufflování na deque je kvadratický, protože neni random access

Comment (jindrahelcl) [reply neuralmonkey/dataset.py]: Navíc nevidim, proč jsem zvolil zrovna `deque`, když na tom volám jen `popleft`, `shuffle`, `extend`, `len` a testuju na prázdnotu

Comment (jindrahelcl) [reply neuralmonkey/dataset.py]: ale jo, když se neshuffluje tak to tam chceš rvát jednou stranou a druhou to házet ven

Comment (jindrahelcl) [reply neuralmonkey/dataset.py]: So, I propose:

1. Instead of a  warning, raise an exception, that the buffer is too small.
2. Document that in the docstring
3. Introduce another option `buffer_min_size`, that would be by default half the size of the buffer. (Think buffer_size of 1000 sentences, after 10 batches of 50 sentences each which we take from the buffer, the buffer gets refilled and reshuffled.)

Comment (jindrahelcl) [reply neuralmonkey/dataset.py]: Anebo 1 škrtnout a přidat jen dokumentaci.

File: neuralmonkey/dataset.py [hunk 9]
```diff
@@ -0,0 +1,559 @@
...
+        if self.buffer_size is not None:
+            # pylint: disable=stop-iteration-return
+            # This is pylint issue https://github.com/PyCQA/pylint/issues/2158
+            buf = deque(next(zipped_iterator) for _ in range(self.buffer_size))
+            # pylint: enable=stop-iteration-return
+        else:
+            assert not self.lazy
+            buf = deque(zipped_iterator)
+        if self.shuffled:
+            random.shuffle(buf)  # type: ignore
```
Comment (jlibovicky) [inline neuralmonkey/dataset.py]: Není efektivnější to lineárně převést na list, pak to lineárně zashuflovat a udělat z toho zase deque? Nebo je se to děje vevnitři v tom shufflu?

File: neuralmonkey/dataset.py [hunk 10]
```diff
@@ -0,0 +1,559 @@
...
+
+
+def from_files(*args, **kwargs):
+    return load_dataset_from_files(*args, **kwargs)
+
+
+# pylint: disable=too-many-locals,too-many-branches
+def load(name: str,
+         series: List[str],
+         sources: List[SourceSpec],
```
Comment (jlibovicky) [inline neuralmonkey/dataset.py]: Nelíbí se mi jméno `sources` protože koliduje se source a target v MT.

File: neuralmonkey/dataset.py (hunk 9 above)
Comment (jindrahelcl) [reply neuralmonkey/dataset.py]: Neděje. Viz co jsem napsal k Dušanovo review. Nejradši bych tam dal minimum size, pod kterou když to vleze, tak se to zashuffluje - klidně převedením na list a zpět. Ale nebude se to shufflovat po každý batchi

File: neuralmonkey/dataset.py (hunk 10 above)
Comment (jindrahelcl) [reply neuralmonkey/dataset.py]: A co se ti líbí?

Comment (jindrahelcl) [reply neuralmonkey/dataset.py]: `data`?

Comment (jlibovicky) [reply neuralmonkey/dataset.py]: Třeba nebo paths

Comment (jindrahelcl) [reply neuralmonkey/dataset.py]: `paths` je blbý protože tam patří i series-level a dataset-level preprocesory. Proto se mi ani nelíbí data.

Comment (jindrahelcl) [reply neuralmonkey/dataset.py]: Ale data docela jde

Comment (varisd) [reply neuralmonkey/dataset.py]: dej tam series_source

Comment (jindrahelcl) [reply neuralmonkey/dataset.py]: no, taky tam můžu dát `víčka_od_škraloupů`

Comment (jlibovicky) [reply neuralmonkey/dataset.py]: Ale data mi pořád přijde lepší než sources.

Comment (jindrahelcl) [reply neuralmonkey/dataset.py]: Ok, dám tam data

Comment (jindrahelcl) [reply neuralmonkey/dataset.py]: hotovo

File: neuralmonkey/dataset.py (hunk 8 above)
Comment (jindrahelcl) [reply neuralmonkey/dataset.py]: htoovo

File: neuralmonkey/dataset.py (hunk 9 above)
Comment (jindrahelcl) [reply neuralmonkey/dataset.py]: snad hotovo. uvidime jestli projdou

Comment (jindrahelcl) [reply neuralmonkey/dataset.py]: tests

**Comment by jindrahelcl** [comment]
I fixed the errors and removed the redundant `lazy` parameter. Now, everytime the `buffer_size` is specified, the dataset will behave lazily (předtim to spadlo, když lazy nebylo true a naopak když lazy bylo true a nebyl specifikovanej buffer size)

**Comment by jindrahelcl** [comment]
Note that the data are actually stored in memory and are not re-read from the files, if the buffer_size is None. This is the expected behavior.

File: neuralmonkey/config/parsing.py [hunk 11]
```diff
@@ -130,10 +130,12 @@ def _parse_list(string: str, vars_dict: VarsDict) -> List[Any]:
...
     values = [_parse_value(val, vars_dict) for val in items]
-    types = [type(val) for val in values]
 
-    if len(set(types)) > 1:
-        raise ParseError("List must of a same type, is: {}".format(types))
+    # Do not check types because the dataset specs list are not homogeneous
+    # types = [type(val) for val in values]
+
+    # if len(set(types)) > 1:
+    #     raise ParseError("List must of a same type, is: {}".format(types))
```
Comment (jlibovicky) [inline neuralmonkey/config/parsing.py]: Proč to celý nesmažeš?

File: neuralmonkey/dataset.py [hunk 12]
```diff
@@ -0,0 +1,572 @@
...
+
+# OutputSpec: Tuple of series name, path, and optionally a writer
+OutputSpec = Union[Tuple[str, str], Tuple[str, str, Writer]]
+# pylint: enable=invalid-name
+
+PREPROCESSED_SERIES = re.compile("pre_([^_]*)$")
+SERIES_SOURCE = re.compile("s_([^_]*)$")
+SERIES_OUTPUT = re.compile("s_(.*)_out")
+
+
```
Comment (jlibovicky) [inline neuralmonkey/dataset.py]: Tady bych možná přidal organizační komentář, že následují věci, co jenom nějak šupačí věci z konfiguráku.

File: neuralmonkey/config/parsing.py (hunk 11 above)
Comment (jindrahelcl) [reply neuralmonkey/config/parsing.py]: Aby sis toho všimnul při review a kdyžtak řek že to nechcem rušit.

Comment (jlibovicky) [reply neuralmonkey/config/parsing.py]: Od tý doby, co jedeme typeguardy, tak už to není potřeba, ne?

Comment (jindrahelcl) [reply neuralmonkey/config/parsing.py]: řek bych že ne; smažu to

File: neuralmonkey/dataset.py [hunk 13]
```diff
@@ -0,0 +1,572 @@
...
+        lazy: bool = False,
+        preprocessors: List[Tuple[str, str, Callable]] = None,
+        **kwargs) -> "Dataset":
+    """Load a dataset from the files specified by the provided arguments.
+
+    Paths to the data are provided in a form of dictionary.
+
+    Keyword arguments:
+        name: The name of the dataset to use. If None (default), the name will
+              be inferred from the file names.
```
Comment (varisd) [inline neuralmonkey/dataset.py]: Tady to chce update docstringu. Z argumentu metody nevypliva, ze None je default + name se neinferuje s filenames, ale z nazvu objektu v configu, ne?

File: neuralmonkey/dataset.py [hunk 14]
```diff
@@ -0,0 +1,572 @@
...
+        lazy: Boolean flag specifying whether to use lazy loading (useful for
+              large files). Note that the lazy dataset cannot be shuffled.
+              Defaults to False.
+        preprocessor: A callable used for preprocessing of the input sentences.
```
Comment (varisd) [inline neuralmonkey/dataset.py]: Tady se nemusi jednat pouze o vety, ne?

Comment (jindrahelcl) [reply neuralmonkey/dataset.py]: na to kašli, to je deprecated funkce

File: neuralmonkey/dataset.py (hunk 13 above)
Comment (jindrahelcl) [reply neuralmonkey/dataset.py]: na to kašli, to je deprecated funkce
